### PR TITLE
Add background_removal tool (skimage)

### PR DESCRIPTION
**Pull request by rmassei**
Hi!
I think it would be nice to have in the collection a tool to automatically perform background removal from an intensity image. 
This tool is based on skimage and more info can be found [here](https://haesleinhuepf.github.io/BioImageAnalysisNotebooks/18_image_filtering/03_background_removal.html).

At the present, three potential options:
1. [Rolling-Ball Algorithm](https://scikit-image.org/docs/stable/auto_examples/segmentation/plot_rolling_ball.html)
2. [Difference of Gaussians](https://scikit-image.org/docs/stable/auto_examples/filters/plot_dog.html)
3. [Top-Hat Filter](https://scikit-image.org/docs/dev/auto_examples/applications/plot_morphology.html#white-tophat)

**Original**
img
**Rolling-Ball**
rolling
**Difference of Gaussians**
gaussian

~~Output is a 3 channel, 16 bit TIFF~~

---

**FOR THE CONTRIBUTOR — Please fill out if applicable**

Please make sure you have read the [CONTRIBUTING.md](https://github.com/BMCV/galaxy-image-analysis/blob/master/CONTRIBUTING.md) document (last updated: 2024/04/23).

* [X] License permits unrestricted use (educational + commercial).

If this PR adds or updates a tool or tool collection:

* [X] This PR adds a new tool or tool collection.
* [ ] This PR updates an existing tool or tool collection.
* [X] Tools added/updated by this PR comply with the [Naming and Annotation Conventions for Tools in the Image Community in Galaxy](https://doi.org/10.37044/osf.io/w8dsz) (or explain why they do not).

### CODE REVIEW

File: tools/background_removal/background_removal.xml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<tool id="background_removal" name="Background Removal" version="0.1">
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:1]:
```suggestion
<tool id="background_removal" name="Remove image background" version="0.1">
```
This is more in line with [our guidelines](https://osf.io/preprints/biohackrxiv/w8dsz)
> To maintain consistency, we consider phrasing names as imperatives a good practice, such as “Analyze particles” or “Perform segmentation using watershed transformation”.

Comment (kostrykin) [inline tools/background_removal/background_removal.xml:1]:
```suggestion
<tool id="background_removal" name="Background Removal" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="20.05">
```
The versioning scheme is suggested by the IUC: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

File: tools/background_removal/background_removal.xml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    <description>with scikit-image</description>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:2]:
```suggestion
    <description>with scikit-image</description>
    <macros>
        <token name="@TOOL_VERSION@">0.24.0</token>
        <token name="@VERSION_SUFFIX@">0</token>
    </macros>
```

File: tools/background_removal/background_removal.xml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        <param name="input_image" type="data" format="tiff, jpg, jpeg, png, tif" label="Input Image"/>
+        <param name="filter" type="select" label="Select Filter">
+            <option value="rolling_ball">Rolling-Ball Algorithm</option>
+            <option value="dog">Difference of Gaussians</option>
+            <option value="top_hat">Top-Hat Filter</option>
+        </param>
+        <param name="radius" type="float" label="Radius" value="20"/>
+    </inputs>
+    <outputs>
+        <data name="output" format="tiff" label="Background Substraction Output"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:21]:
```suggestion
        <data name="output" format="tiff" label="Background substraction output"/>
```
This is more consistent with the other tools.

File: tools/background_removal/background_removal.xml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    </outputs>
+    <tests>
+        <test>
+            <param name="input_image" value="img.png"/>
+            <param name="filter" value="rolling_ball"/>
+            <param name="radius" value="20"/>
+            <output name="output" value="out_rolling_ball.tiff" ftype="tiff"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:28]: Can you please use a suitable compare method for the tests, like `image_diff`?
https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output

Alternatively, I'd suggest using one of the macros we have pre-defined for images. See this for example:
https://github.com/BMCV/galaxy-image-analysis/blob/5a1ff1a3a5df15f75c95734af3ead420de536c0b/tools/2d_simple_filter/filter.xml#L116

File: tools/background_removal/background_removal.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        This tools applied different background removal algorithm to the image:
+
+        - Rolling-Ball Algorithm
+
+        - Difference of Gaussians
+
+        - Top-Hat Filter
+    </help>
+    <citations>
+        <citation type="doi">https://haesleinhuepf.github.io/BioImageAnalysisNotebooks/18_image_filtering/03_background_removal.html</citation>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:53]: I think that a URL is not a valid DOI, is it?

I think you should cite [DOI:10.1109/MC.1983.1654163](https://doi.org/10.1109/MC.1983.1654163) here, and maybe some respective reference from https://en.wikipedia.org/wiki/Top-hat_transform#References, if there is one.

**Review by kostrykin — CHANGES_REQUESTED** [review]
Thanks a lot for this work!

There are a few comments below.

Also, please feel free to add yourself as a creator, as in:
https://github.com/BMCV/galaxy-image-analysis/blob/5a1ff1a3a5df15f75c95734af3ead420de536c0b/tools/bfconvert/bfconvert.xml#L10

**Comment by bgruening** [comment]
> Input staging problem: Test input file (img.png) cannot be found.

You can find the planemo output on this site for example: https://github.com/BMCV/galaxy-image-analysis/actions/runs/9907411219

File: tools/background_removal/background_removal.xml (hunk 2 above)
Comment (rmassei) [reply tools/background_removal/background_removal.xml]: Done!

File: tools/background_removal/background_removal.xml (hunk 3 above)
Comment (rmassei) [reply tools/background_removal/background_removal.xml]: Done!

File: tools/background_removal/background_removal.xml (hunk 4 above)
Comment (rmassei) [reply tools/background_removal/background_removal.xml]: I just added the macro tested locally and everything seems to work. Let me know if this is done correctly, hope I understood it right

File: tools/background_removal/background_removal.xml (hunk 5 above)
Comment (rmassei) [reply tools/background_removal/background_removal.xml]: Added just the doi, I am not so sure how to add correctly the wikipedia reference...

**Review by rmassei — COMMENTED** [review]
Accepted all the changes and added the macro tests

File: tools/background_removal/background_removal.xml (hunk 1 above)
Comment (rmassei) [reply tools/background_removal/background_removal.xml]: Done

File: tools/background_removal/background_removal.xml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    </inputs>
+    <outputs>
+        <data name="output" format="tiff" label="Background substraction output"/>
+    </outputs>
+    <tests>
+        <test>
+            <param name="input_image" value="img.jpg"/>
+            <param name="filter" value="rolling_ball"/>
+            <param name="radius" value="20"/>
+            <expand macro="tests/intensity_image_diff" name="output" value="img_output_rb.tif" ftype="tiff"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:37]: There is no such file `img_output_rb.tif`?

You probably meant `out_rolling_ball.tif`?
```suggestion
            <expand macro="tests/intensity_image_diff" name="output" value="out_rolling_ball.tif" ftype="tiff"/>
```

However, I noticed that `img.jpg` is an 8bit image, but `out_rolling_ball.tif` is 16bits. As a consequence, the range of values is not preserved. This is unconvenient in practice, [right](https://github.com/BMCV/galaxy-image-analysis/issues/105#issuecomment-1994180241)?

**Review by kostrykin — COMMENTED** [review]
Thanks @rmassei!

There are a few more comments below.

File: tools/background_removal/background_removal.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        output_image = image - background_rolling
+
+    elif args.filter == "dog":
+        output_image = difference_of_gaussians(image, low_sigma=0, high_sigma=args.radius)
+
+    elif args.filter == "top_hat":
+        output_image = white_tophat(image, disk(args.radius))
+
+    with warnings.catch_warnings():
+        output_image = skimage.util.img_as_uint(output_image)
```
Comment (kostrykin) [inline tools/background_removal/background_removal.py:26]:
```suggestion
        output_image = convert_image_to_format_of(output_image, image)
```
and then add a function like:
```python
def convert_image_to_format_of(image, format_image):
    """
    Converts the first image to the format of the second image.
    """
    if format_image.dtype == image.dtype:
        return image
    elif format_image.dtype == np.uint8:
        return skimage.util.img_as_ubyte(image)
    elif format_image.dtype == np.uint16:
        return skimage.util.img_as_uint(image)
    elif format_image.dtype == np.int16:
        return skimage.util.img_as_int(image)
    else:
        raise ValueError(f'Unsupported image data type: {format_image.dtype}')
```

File: tools/background_removal/background_removal.xml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            <option value="top_hat">Top-Hat Filter</option>
+        </param>
+        <param name="radius" type="float" label="Radius" value="20"/>
+    </inputs>
+    <outputs>
+        <data name="output" format="tiff" label="Background substraction output"/>
+    </outputs>
+    <tests>
+        <test>
+            <param name="input_image" value="img.tif"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:34]: This is an uint16 image now, right? Then, please rename `img.tif` to `img_uint16.tif` and
```suggestion
            <param name="input_image" value="img_uint16.tif"/>
```
Also for the other tests below.

File: tools/background_removal/background_removal.xml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            <option value="dog">Difference of Gaussians</option>
+            <option value="top_hat">Top-Hat Filter</option>
+        </param>
+        <param name="radius" type="float" label="Radius" value="20"/>
+    </inputs>
+    <outputs>
+        <data name="output" format="tiff" label="Background substraction output"/>
+    </outputs>
+    <tests>
+        <test>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:33]: The `img.tif` file is an uint16 image now, right?

I suggest renaming it to `input2_uint16.tif` then. In addition, please also add the uint8 variant back, and name it `input1_uint8.tif`, so we can do the following:

```suggestion
        <!-- uint8 tests -->
        <test>
            <param name="input_image" value="input1_uint8.tif"/>
            <param name="filter" value="dog"/>
            <param name="radius" value="20"/>
            <expand macro="tests/intensity_image_diff" name="output" value="input1_output_dog.tif" ftype="tiff"/>
        </test>
        <!-- uint16 tests -->
        <test>
```

File: tools/background_removal/background_removal.xml (hunk 8 above)
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:34]:
```suggestion
            <param name="input_image" value="input2_uint16.tif"/>
```

File: tools/background_removal/background_removal.xml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            <param name="input_image" value="img.tif"/>
+            <param name="filter" value="rolling_ball"/>
+            <param name="radius" value="20"/>
+            <expand macro="tests/intensity_image_diff" name="output" value="img_output_rb.tif" ftype="tiff"/>
+        </test>
+        <test>
+            <param name="input_image" value="img.tif"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:40]:
```suggestion
            <param name="input_image" value="input2_uint16.tif"/>
```

File: tools/background_removal/background_removal.xml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            <param name="filter" value="dog"/>
+            <param name="radius" value="20"/>
+            <expand macro="tests/intensity_image_diff" name="output" value="img_output_dog.tif" ftype="tiff"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:43]:
```suggestion
            <expand macro="tests/intensity_image_diff" name="output" value="input2_output_dog.tif" ftype="tiff"/>
```

File: tools/background_removal/background_removal.xml [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        </test>
+        <test>
+            <param name="input_image" value="sample.tif"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:46]: Maybe rename `sample.tif` to `input3_uint8.tif` and then:
```suggestion
            <param name="input_image" value="input3_uint8.tif"/>
```

File: tools/background_removal/background_removal.xml [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            <param name="filter" value="top_hat"/>
+            <param name="radius" value="15"/>
+            <expand macro="tests/intensity_image_diff" name="output" value="img_output_tophat.tif" ftype="tiff"/>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:49]:
```suggestion
            <expand macro="tests/intensity_image_diff" name="output" value="input3_tophat.tif" ftype="tiff"/>
```

**Review by kostrykin — COMMENTED** [review]
So I think there was a misunderstanding about my previous comment :)
https://github.com/BMCV/galaxy-image-analysis/pull/125#discussion_r1677083102

I mean, the issue that the range of values is changed by the tool remains, even after you changed the test input data to an uint16 image. We need to dig a bit deeper here.

First, lets see how the wrapped functions change the data type:
- [skimage.filters.difference_of_gaussians](https://scikit-image.org/docs/stable/api/skimage.filters.html#skimage.filters.difference_of_gaussians) changes everything to float.
- [skimage.morphology.white_tophat](https://scikit-image.org/docs/stable/api/skimage.morphology.html#skimage.morphology.white_tophat) does not change the data type.
- [skimage.restoration.rolling_ball](https://scikit-image.org/docs/stable/api/skimage.restoration.html#skimage.restoration.rolling_ball) does not change the data type either.

So we only need to care for conversion to float. I suggest something like in my review, but maybe it can be done more elegantly.

**Comment by rmassei** [comment]
thanks @kostrykin, I did not get the problem in the first place.
Test were updated and it seems that the new function is working by correctly converting to the input data type.

File: tools/background_removal/background_removal.xml [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            <param name="filter" value="dog"/>
+            <param name="radius" value="20"/>
+            <expand macro="tests/intensity_image_diff" name="output" value="input2_output_dog.tif" ftype="tiff"/>
+        </test>
+        <test>
+            <param name="input_image" value="input3_uint8.tif"/>
+            <param name="filter" value="top_hat"/>
+            <param name="radius" value="15"/>
+            <expand macro="tests/intensity_image_diff" name="output" value="input3_output_tophat.tif" ftype="tiff"/>
+        </test>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:58]:
```suggestion
```
We should move this up to the uint8 test section, right?

File: tools/background_removal/background_removal.xml [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        <data name="output" format="tiff" label="Background substraction output"/>
+    </outputs>
+    <tests>
+        <!-- uint8 tests -->  
+        <test>  
+            <param name="input_image" value="input1_uint8.tif"/>  
+            <param name="filter" value="dog"/>  
+            <param name="radius" value="20"/>  
+            <expand macro="tests/intensity_image_diff" name="output" value="input1_output_dog.tif" ftype="tiff"/>  
+        </test>
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:39]:
```suggestion
        </test>
        <test>
            <param name="input_image" value="input3_uint8.tif"/>
            <param name="filter" value="top_hat"/>
            <param name="radius" value="15"/>
            <expand macro="tests/intensity_image_diff" name="output" value="input3_output_tophat.tif" ftype="tiff"/>
        </test>
```

File: tools/background_removal/background_removal.xml [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    </tests>
+    <help>
+        This tools applied different background removal algorithm to the image:
+
+        - Rolling-Ball Algorithm
+
+        - Difference of Gaussians
+
+        - Top-Hat Filter
```
Comment (kostrykin) [inline tools/background_removal/background_removal.xml:67]:
```suggestion
        This tool applies different background removal algorithms to an image:

        - Rolling-Ball algorithm

        - Difference of Gaussians

        - Top-Hat filter
```
This fixes some spelling / capitalization (Difference of Gaussians is a proper name, but algorithm isn't, and neither is filter).

**Review by kostrykin — COMMENTED** [review]
Thanks @rmassei, I think we are almost done :) only a few minor comments above.

And there are a few flake8 linting issues which need to be addressed:
https://github.com/BMCV/galaxy-image-analysis/actions/runs/9937226470/job/27448034390?pr=125#step:6:1

> tools/background_removal/background_removal.py:28:1: W293 blank line contains whitespace
tools/background_removal/background_removal.py:29:1: E302 expected 2 blank lines, found 1
tools/background_removal/background_removal.py:29:53: W291 trailing whitespace
tools/background_removal/background_removal.py:30:8: W291 trailing whitespace
tools/background_removal/background_removal.py:31:63: W291 trailing whitespace
tools/background_removal/background_removal.py:32:8: W291 trailing whitespace
tools/background_removal/background_removal.py:33:42: W291 trailing whitespace
tools/background_removal/background_removal.py:34:21: W291 trailing whitespace
tools/background_removal/background_removal.py:35:41: W291 trailing whitespace
tools/background_removal/background_removal.py:36:48: W291 trailing whitespace
tools/background_removal/background_removal.py:37:42: W291 trailing whitespace
tools/background_removal/background_removal.py:38:47: W291 trailing whitespace
tools/background_removal/background_removal.py:39:41: W291 trailing whitespace
tools/background_removal/background_removal.py:40:46: W291 trailing whitespace
tools/background_removal/background_removal.py:41:10: W291 trailing whitespace
tools/background_removal/background_removal.py:42:79: W291 trailing whitespace

File: tools/background_removal/background_removal.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    elif args.filter == "dog":
+        output_image = difference_of_gaussians(image, low_sigma=0, high_sigma=args.radius)
+
+    elif args.filter == "top_hat":
+        output_image = white_tophat(image, disk(args.radius))
+
+    with warnings.catch_warnings():
+        output_image = convert_image_to_format_of(output_image, image)
+        skimage.io.imsave(args.output, output_image, plugin="tifffile")
+
```
Comment (kostrykin) [inline tools/background_removal/background_removal.py:28]: @rmassei one flake8 error remained:
> tools/background_removal/background_removal.py:29:1: E302 expected 2 blank lines, found 1
https://github.com/BMCV/galaxy-image-analysis/actions/runs/9937768191/job/27448897469?pr=125#step:6:16
```suggestion


```

**Review by kostrykin — APPROVED** [review]
Much thanks!

**Comment by kostrykin** [comment]
https://github.com/BMCV/galaxy-image-analysis/actions/runs/9937821722?pr=125
> pillow and tifffile are not installed, but required to compare files using the 'image_diff' method

I think this is due to https://github.com/BMCV/galaxy-image-analysis/pull/124
where we removed `python -m pip install pillow tifffile` in https://github.com/BMCV/galaxy-image-analysis/pull/124/commits/6bc8575d700f6ec7a194566a335c4879385f12b4

@bernt-matthias Any idea how to fix this? Shall we add it back again?

**Comment by rmassei** [comment]
yes, I got the same error locally before running `python -m pip install pillow tifffile`

**Comment by kostrykin** [comment]
> Could not update background_removal
Failed to find repository for owner/name ufz/background_removal
Repository background_removal updated successfully.
Could not update background_removal
Failed to find repository for owner/name ufz/background_removal
Unexpected HTTP status code: 400: {"err_msg": "You already own a repository named <b>background_removal</b>, please choose a different name.", "err_code": 400008}
Repository [background_removal] does not exist in the targeted Tool Shed.
Failed to update metadata for repository 'background_removal' on the main Tool Shed.
Repository contents updated but failed to update metadata for repository 'background_removal' on the main Tool Shed.

https://github.com/BMCV/galaxy-image-analysis/actions/runs/9946388590/job/27477147950#step:6:222

Not sure what this error means. @bgruening do you know?

On the one hand it says that `background_removal` repository is already owned, on the other hand it complains that it doesn't exist… shouldn't it create it?

File: tools/background_removal/.shed.yml [hunk 18]
```diff
@@ -0,0 +1,8 @@
+categories:
+  - Imaging
+description: Background removal filters using scikit-image
+long_description: Tool to perform a background removal using 1) Rolling-Ball Algorithm, 2) Difference of Gaussians and 3) Top-Hat Filter
+name: background_removal
+owner: ufz
```
Comment (bgruening) [inline tools/background_removal/.shed.yml:6]: the owner is wrong, we can only push to imgteam

**Comment by bgruening** [comment]
Please change the owner in the shed.yml file

**Pull request merged**